lib/src/main/java/graphql/nadel/engine/transform/partition/NadelPartitionTransform.kt

```diff
@@ -0,0 +1,266 @@
+            fieldPartitionContext = fieldPartitionContext,
+            graphQLSchema = executionBlueprint.engineSchema,
+            partitionKeyExtractor = partitionTransformHook.getPartitionKeyExtractor()
+        )
+
+        val fieldPartitions = try {
+            fieldPartition.createFieldPartitions(field = overallField)
+        } catch (exception: Exception) {
+            return State(
+                executionContext = executionContext,
+                fieldPartitionContext = fieldPartitionContext,
+                errors = mutableListOf(exception)
+            )
+        }
+
+        if (fieldPartitions.size < 2) {
+            // We can't partition a field that doesn't have at least two partitions
+            return null
+        }
+
+        partitionTransformHook.willPartitionCallback(executionContext, fieldPartitions)
+
+        return State(
+            executionContext = executionContext,
+            fieldPartitionContext = fieldPartitionContext,
+            fieldPartitions = fieldPartitions
+        )
+    }
+
+    private fun getPathToPartitionArg(overallField: ExecutableNormalizedField, schema: GraphQLSchema): List<String>?  {
+        val fieldDef = overallField.getFieldDefinitions(schema).first()
+        val routingDirective = fieldDef.getAppliedDirective(NadelDirectives.partitionDirectiveDefinition.name) ?: return null
+        val pathToSplitPoint = routingDirective.getArgument("pathToSplitPoint")
+            ?.argumentValue
+            ?.value as ArrayValue?
+
+        return pathToSplitPoint?.values?.map { it as StringValue }?.map { it.value }
+    }
+
+    override suspend fun transformField(
+        executionContext: NadelExecutionContext,
+        serviceExecutionContext: NadelServiceExecutionContext,
+        transformer: NadelQueryTransformer,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        service: Service,
+        field: ExecutableNormalizedField,
+        state: State,
+    ): NadelTransformFieldResult {
+
+        if (state.errors != null) {
+            // At this point we know we can't partition the field, but removing it could result
+            // in a query with no top-level fields, which would be invalid.
+            // Adding __typename as an artificial field also wouldn't work because that field
+            // belongs to all operation types, which causes issues further down the execution.
+            return NadelTransformFieldResult.unmodified(field)
+        }
+
+        val fieldPartitions = checkNotNull(state.fieldPartitions) { "Expected fieldPartitions to be set" }
+
+        val primaryPartition = fieldPartitions.values.first()
+
+        // TODO: throw error if operation is Subscription?
+        val rootType = executionContext.query.operation.getType(executionBlueprint.engineSchema)
+
+        val partitionCalls = coroutineScope {
+            fieldPartitions.values.drop(1).map {
```

Comment:
Could put `fieldPartitions.values.drop(1)` right below the definition of `primaryPartition` so it's easier to understand i.e.

```kotlin
val primaryPartition = fieldPartitions.values.first()
val otherPartitions = fieldPartitions.values.drop(1)

```

Reply:
Nice